spanner/spanner_samples.rb

```diff
@@ -0,0 +1,390 @@
+    end
+
+    new_first_album_budget  = first_album[:MarketingBudget]  + 200_000
+    new_second_album_budget = second_album[:MarketingBudget] - 200_000
+
+    transaction.update "Albums", [
+      { SingerId: 1, AlbumId: 1, MarketingBudget: new_first_album_budget  },
+      { SingerId: 2, AlbumId: 2, MarketingBudget: new_second_album_budget }
+    ]
+  end
+
+  puts "Transaction complete"
+  # [END read_write_transaction]
+end
+
+def query_data_with_index project_id:, instance_id:, database_id:
+  # [START query_data_with_index]
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  sql_query = 'SELECT AlbumId, AlbumTitle, MarketingBudget
+               FROM Albums@{FORCE_INDEX=AlbumsByAlbumTitle}
+               WHERE AlbumTitle >= "Aardvark" AND AlbumTitle < "Goo"'
+
+  client.execute(sql_query).rows.each do |row|
+    puts "#{row[:AlbumId]} #{row[:AlbumTitle]} #{row[:MarketingBudget]}"
+  end
+  # [END query_data_with_index]
+end
+
+def read_data_with_index project_id:, instance_id:, database_id:
+  # [START read_data_with_index]
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  result = client.read "Albums", [:AlbumId, :AlbumTitle],
+                       index: "AlbumsByAlbumTitle"
+
+  result.rows.each do |row|
+    puts "#{row[:AlbumId]} #{row[:AlbumTitle]}"
+  end
+  # [END read_data_with_index]
+end
+
+def read_data_with_storing_index project_id:, instance_id:, database_id:
+  # [START read_data_with_storing_index]
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  result = client.read "Albums", [:AlbumId, :AlbumTitle, :MarketingBudget],
+                       index: "AlbumsByAlbumTitle2"
+
+  result.rows.each do |row|
+    puts "#{row[:AlbumId]} #{row[:AlbumTitle]} #{row[:MarketingBudget]}"
+  end
+  # [END read_data_with_storing_index]
+end
+
+def read_only_transaction project_id:, instance_id:, database_id:
+  # [START read_only_transaction]
+  # project_id  = "Your Google Cloud project ID"
+  # instance_id = "Your Spanner instance ID"
+  # database_id = "Your Spanner database ID"
+
+  require "google/cloud/spanner"
+
+  spanner = Google::Cloud::Spanner.new project: project_id
+  client  = spanner.client instance_id, database_id
+
+  result = client.read "Albums", [:AlbumId, :AlbumTitle, :SingerId],
```

Comment:
Why not use [snapshot](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud/v0.35.0/google/cloud/spanner/client?method=snapshot-instance)? IIRC, you mentioned other languages use this higher level method for read only transactions. Might make this sample easier to read!